python-client/giskard/datasets/base/__init__.py

```diff
@@ -315,39 +312,39 @@ def _infer_column_types(self, cat_columns: List[str]):
         Returns:
             dict: A dictionary that maps column names to their inferred types, one of 'text', 'numeric', or 'category'.
         """
-        column_types = {}
-        nuniques = self.df.nunique()
-        df_columns = list(self.df.columns)
+        _column_types = column_types.copy()
+        df_columns = set(self.df.drop([self.target], axis=1).columns) if self.target else set(self.df.columns)
 
+        # priority of cat_columns over column_types (for categorical columns)
         if cat_columns:
             if not set(cat_columns).issubset(df_columns):
                 raise ValueError(
                     "The provided 'cat_columns' are not all part of your dataset 'columns'. "
                     "Please make sure that `cat_columns` refers to existing columns in your dataset."
                 )
-
             for cat_col in cat_columns:
                 if cat_col != self.target:
-                    column_types[cat_col] = SupportedColumnTypes.CATEGORY.value
-            df_columns = set(df_columns) - set(cat_columns)
+                    _column_types[cat_col] = SupportedColumnTypes.CATEGORY.value
```

Comment:
Maybe more concise
```py
_column_types.update({col: SupportedColumnTypes.CATEGORY for col in cat_columns})
```

Reply:
Should we check if `cat_columns` are compatible with the `column_types` declared?